Adding DiagnosticSource to EventSource Bridge.

vancem opened the pull request:
Currently DiagnosticSource just goes to DiagnosticListeners.   We want to
be able to have the capabilty to pipe DiagnosticSource information to
EventSource consumers (ETW and EventListeners).  We do this with a new
DiagnosticSource EventSource called 'Microsoft-DiagnosticsSource'  If
EventListener / ETW subscribe to that EventSource they can get events
from any DiangnosticSouce.

See the comments in the code itself as well as the tests in the file
DiagnosticSourceEventSourceBridgeTests.cs for more on specifics of
using this new functionality.

vancem commented:
@stephentoub, @ericstj, @weshaggard 

The changes to the project.json and the .csproj and .builds file are mostly because I want this package to work on V4.5 systems as well as .NET Core.   Sadly V4.5 runtimes don't have the newer ability to send complex types (e.g. arrays of key value pairs), in the eventsource payload.   Thus the package has separate code for older runtimes in which the output is sent as a single JSON stream for those cases (it is the best we can do).   It is not clear how important the V4.5 case is, so we may drop it in the relatively near future, but it does work right now if necessary.

Also note that no new APIs were introduced.   This works with existing EventSource mechanisms.  Also the EventSource we are using already existed (but did relatively little).   Now it has forwarding capability.  

There are know efficiency issues if this bridge is used extensively, but it seemed wise to wait until we have a better idea of how this bridge is likely to be used in practice before optimizing.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSource.cs:
```diff
@@ -18,6 +19,7 @@ public partial class DiagnosticListener : System.Diagnostics.DiagnosticSource, S
   }
   public abstract partial class DiagnosticSource {
     protected DiagnosticSource() { }
+    public static System.Diagnostics.DiagnosticSource DefaultSource { get { return default(System.Diagnostics.DiagnosticSource); } }
```
Didn't we just remove both of these in https://github.com/dotnet/corefx/commit/8a39134f5b6058ba4ec1b01153ee78597b9ba295?  Are they being added back on purpose?

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.builds:
```diff
@@ -2,8 +2,10 @@
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <ItemGroup>
-    <Project Include="System.Diagnostics.DiagnosticSource.csproj" />
+    <Project Include="System.Diagnostics.DiagnosticSource.csproj"/>
```
Is a ```<TargetGroup/>``` needed for this first one, too?

vancem replied:
The .csproj file needs to work without wrapping '.builds' wrapper, so you pick the most interesting build variation to build by default (e.g. in VS).   This happens to be netstandard1.3.   Thus the first one is the 'default' one and does not need a TargetGroup.    I could put an explicit TargetGroup here, or a comment, but that has its own potential for mistakes.   I am following conventions used elsewhere when I am not specifying the targetGroup for the first build in the builds file, so I think it is OK as it is...

weshaggard replied:
Agreed I wouldn't add a TargetGroup for this first entry which is an empty/default TargetGroup.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticListener.cs:
```diff
@@ -34,6 +34,16 @@ namespace System.Diagnostics
+        /// events in the system (e.g. multi-tenancy).  
+        /// </summary>
+        public static DiagnosticListener DefaultListener { get { return s_default; } }
+#endif 
```
Same question about these ```#if false``` blocks... they were just removed in https://github.com/dotnet/corefx/commit/8a39134f5b6058ba4ec1b01153ee78597b9ba295... why are they being added back?

vancem replied:
This was a merge error.  Fixed.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -2,32 +2,193 @@
+    ///   what DiagnsoticSources to enable and what parts of the payload to serialize into the key-value
+    ///   list that will be forwarded to the EventSource.    If it is empty, all serializable parts of
+    ///   every DiagnosticSource event will be forwarded (this is NOT recommended for monitoring but 
+    ///   can be useful for discover).
```
Should this be "discovery" rather than "discover"?

vancem replied:
Fixed.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -2,32 +2,193 @@
+    ///   every DiagnosticSource event will be forwarded (this is NOT recommended for monitoring but 
+    ///   can be useful for discover).
+    ///   
+    ///   The FilterAndPayloadSpecs is one long string with a the following structures
```
Nit: remove "a" in "a the"

vancem replied:
Fixed.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -2,32 +2,193 @@
+    ///   
+    ///   The FilterAndPayloadSpecs is one long string with a the following structures
+    ///   
+    ///   * It is a newline separated list of FILTER_AND_PAYLOAD_SPEC
```
Newlines change between platforms.  Is there any value in using a separator that doesn't change based on platform?

vancem replied:
The thing I like about newlines is it breaks up what might otherwise be a very long specification when viewed with normal editing tools.   The code will accept both \n and \r\n as the separator, so should work well on both Windows and Linux.    I am open to other options, but I like newline because it does break it up.

stephentoub replied:
That's fine.  I saw later that you handle both ```\r\n``` and ```\n```, so my concerns are mitigated.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -2,32 +2,193 @@
     {
         public static DiagnosticSourceEventSource Logger = new DiagnosticSourceEventSource();
 
+        class Keywords
```
Nit: we try to be explicit about all visibility in the repo, e.g. ```private class Keywords```

vancem replied:
fixed

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -2,32 +2,193 @@
+        /// Used to send ad-hoc diagnostics to humans.   
+        /// </summary>
+        [Event(1, Keywords = Keywords.Messages)]
+        public void Message(string Message)
```
Are you using capitalized argument names on purpose?

vancem replied:
Yes.   The names of arguments end up being the names of properties when the event is parsed.   Thus one way or the the other our casing convention gets violated.   I choose the convention for properties, because that is probably how people think about them most of the time, and that is how you would do it if you use Write<T>( .. new { ...}).

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -2,32 +2,193 @@
+        /// This is only used on V4.5 systems that dont have the ability to log KeyValuePairs directly.
+        /// It will eventually go away, but we should alwasy reserve the ID for this.    
+        /// </summary>
+        [Event(3)]
```
Should this have ```, Keywords = Keywords.Events``` as well?

vancem replied:
Fixed.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+                    startTransformIdx = colonIdx + 1;
+                }
+
+                // Parse the Soruce/Event name into listenerNameFilter and eventNameFilter
```
nit: Soruce => Source

vancem replied:
fixed.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+                    eventNameFilter = filterAndPayloadSpec.Substring(slashIdx + 1, endEventNameIdx - slashIdx - 1);
+                }
+                else if (startIdx < endEventNameIdx)
+                    listenerNameFilter = filterAndPayloadSpec.Substring(startIdx, endEventNameIdx - startIdx);
```
Nit: please use braces (for consistency, since the if block has braces)

vancem replied:
Fixed

KrzysztofCwalina commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -2,32 +2,193 @@
+        private static string ToJSON(List<KeyValuePair<string, string>> keyValues)
         {
+            StringBuilder sb = new StringBuilder();
+            sb.AppendLine("{");
```
This could append '{', i.e. char

vancem replied:
It turns out that AppendLine does not have a character overload, just a string one.   It is also hopefully a non-common code path (only used in V4.5 case), so I did not try to optimize it (e.g. reuse the stringbuilder ...)

KrzysztofCwalina replied:
Interesting. I might add such overload then,

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+                    listenerNameFilter = filterAndPayloadSpec.Substring(startIdx, endEventNameIdx - startIdx);
+
+
+                _eventSource.Message("DiagnosticSource: Enabling '" + (listenerNameFilter ?? "*") + "/" + (eventNameFilter ?? "*") + "'");
```
You might consider using C# 6 interpolated strings for some of these.

KrzysztofCwalina commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -2,32 +2,193 @@
+        /// It will eventually go away, but we should alwasy reserve the ID for this.    
+        /// </summary>
+        [Event(3)]
+        private void EventJSON(string SourceName, string EventName, string ArgmentsJSON)
```
per our naming guidelines, JSON should be cased "Json".

vancem replied:
fixed

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+            Type _expectedArgType;                         // This is the type where 'implicitTransforms is built for'
+            TransformSpec _implicitTransforms;
+            TransformSpec _explicitTransforms;
+            DiagnosticSourceEventSource _eventSource;
```
Nit: explicit visibility for all of these

vancem replied:
fixed

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+            /// and efficienctly.   Thus it represents a '.PROP' in a TransformSpec
+            /// (and a transformSpec has a list of these).  
+            /// </summary>
+            class PropertySpec
```
Nit: explicit visibility

vancem replied:
fixed

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+
+                    #region private 
+
+                    class TypedFetchProperty<Tobject, Tproperty> : PropertyFetch
```
Nit: explicit visibility

stephentoub replied:
Nit: ideally it'd also be ```TObject``` and ```TProperty``` instead of ```Tobject``` and ```Tproperty```

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+            }
+
+            string _outputName;
+            PropertySpec _fetches;
```
Nit: explicit visibility for all of these fields.  I'll stop commenting on these :smile:

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+        {
+            public CallbackObserver(Action<T> callback) { _callback = callback; }
+
+            #region private 
```
What does it mean for public members to be in a ```#region private```?

vancem replied:
This is debatable.    For me, private means 'Clients do not need to (and would prefer not to) see this'  

This includes all privates, but it also includes implementations of contracts that a client is not likely to call directly.   Thus implementations of Dispose() or in this case IObservable all into this category (showing them to clients does not really tell you anything as the public inheritance contract already says it is IDisposable or IObservable, and the fact that there is an explicit implementation is a hidable detail) so I hide it.

The goal is to segregate nicely what a CLIENT wants to see.

stephentoub replied:
It's not a big deal, but if it were me, given that argument, I'd have used explicit interface implementation, e.g.
```C#
void IObserver<T>.OnCompleted() {}
void IObserver<T>.OnError(Exception error) {}
void IObserver<T>.OnNext(T value) { _callback(value); }
```
A "client" of this class doesn't see whether there's a region in the source.

vancem replied:
Yes, that would make it purer.  It does not work on abstract types or overrides (e.g. ToString).  It also has the side effect of making it harder to call the implemented method (you have to cast it to the interface before making the call).   

But mostly I was just trying to do something simple (hide methods the client does not care about), and don't really want to perturb the code just to make it pure.   I could name the region 'non-client' which more accurately describes the partitioning, but without explaination, it is probably no better than 'private'.     

For what it is worth, if push came to shove I would rather simply move the public things out of the private region, but it often does clutter the client interface (since we put thing in the private region because they are more implementation than client contract).

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+        // A linked list of IObservable subscriptions (which are IDisposable).  
+        // We use this to keep track of the DiagnosticSource subscriptions.  
+        // We use this linked list for thread atomicity and to avoid taking a 
+        // dependency on collections.  
```
> to avoid taking a dependency on collections

Yet ```"System.Collections": "4.0.0"``` is being added to the project.json... am I'm misunderstanding?

vancem replied:
Since the comment was written, collections were dragged in.  I have fixed the comment.   Linked lists still do make sense for lists that tend to have only 1 element in them.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -2,32 +2,193 @@
+                    sb.Append(',').AppendLine();
+                first = false;
+
+                // TODO escape quotes in value.   
```
Is there an issue associated with this TODO?

vancem replied:
I have fixed the code to deal with the TODO

justinvp commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+                        {
+                            return _propertyFetch((Tobject)obj);
+                        }
+                        private Func<Tobject, Tproperty> _propertyFetch;
```
Nit: `readonly`

vancem replied:
fixed

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+
+                        var subscription = newListener.Subscribe(new CallbackObserver<KeyValuePair<string, object>>(delegate (KeyValuePair<string, object> evnt)
+                        {
+                            // THe filter is only a best effort here we do it for real.   
```
Nit: THe => The.  I also don't understand the "we do it for real" part.

vancem replied:
Fixed (and better comment)

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+                            {
+                                var propertyType = property.PropertyType;
+                                if (propertyType == typeof(string) || propertyType.GetTypeInfo().IsPrimitive)
+                                    newSerializbleArgs = new TransformSpec(property.Name, 0, property.Name.Length, newSerializbleArgs);
```
If it's some other type, we just ignore the property silently?  Just making sure that's the desired behavior.

vancem replied:
Yes, that is the desired behavior.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+                        {
+                            // Figure out the default properties to send on to EventSource.  These are all string or primitive properties.  
+                            _implicitTransforms = null;
+                            TransformSpec newSerializbleArgs = null;
```
Nit: newSerializbleArgs => newSerializableArgs

vancem replied:
fixed.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/tests/DiagnosticSourceEventSourceBridgeTests.cs:
```diff
@@ -0,0 +1,557 @@
+
+                // Turn on events with both implicit and explicit types 
+                eventSourceListener.Enable(
+                    "BridgeTestSource1/TestEvent1:cls_Point_X=cls.Point.X;cls_Point_Y=cls.Point.Y\r\n" +
```
Can we use Environment.NewLine rather than ```"\r\n"``` in the test, to make sure this actually works correctly with the newlines on unix?

vancem replied:
Sure

stephentoub commented on src/System.Diagnostics.DiagnosticSource/tests/DiagnosticSourceEventSourceBridgeTests.cs:
```diff
@@ -0,0 +1,557 @@
+                Assert.Equal("MyUrl", eventSourceListener.LastEvent.Arguments["cls_Url"]);
+                eventSourceListener.ResetEventCountAndLastEvent();
+
+                // Emit and event that does not match either pattern.  (thus will be filtered out)
```
Nit: and => an

vancem replied:
fixed.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/tests/DiagnosticSourceEventSourceBridgeTests.cs:
```diff
@@ -0,0 +1,557 @@
+                }
+                Assert.Equal(0, eventSourceListener.EventCount);        // No Event should be fired.  
+            }
+            Debug.WriteLine("");    // A place to put a breakpoint.  
```
I'm not understanding this... a place to put a breakpoint in the test?

vancem replied:
removed.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/tests/DiagnosticSourceTests.cs:
```diff
@@ -372,6 +372,12 @@ public void AllListenersAddRemove()
                 DiagnosticListener returnedListener = null;
                 Action<DiagnosticListener> onNewListener = delegate (DiagnosticListener listen)
                 {
+                    // Other tests can be running concurrently with this test, which will make
```
If you want to disable that, you can, e.g. https://github.com/dotnet/corefx/blob/master/src/System.Linq.Parallel/tests/XunitAssemblyAttributes.cs#L10

vancem replied:
Good to know.  Thanks.

stephentoub commented:
Test failure on Ubuntu Release:
```
09:46:11    System.Diagnostics.Tests.DiagnosticSourceEventSourceBridgeTests.TestNulls [FAIL]
09:46:11       Assert.Equal() Failure
09:46:11       Expected: 1
09:46:11       Actual:   2
09:46:11       Stack Trace:
09:46:11 The parameters to the Event method do not match the parameters to the WriteEvent method. This may cause the event to be displayed incorrectly.
09:46:11 The parameters to the Event method do not match the parameters to the WriteEvent method. This may cause the event to be displayed incorrectly.
09:46:11    System.Security.Cryptography.Primitives.Tests  Total: 43, Errors: 0, Failed: 0, Skipped: 0, Time: 1.010s
09:46:11 The parameters to the Event method do not match the parameters to the WriteEvent method. This may cause the event to be displayed incorrectly.
09:46:11 The parameters to the Event method do not match the parameters to the WriteEvent method. This may cause the event to be displayed incorrectly.
09:46:11 The parameters to the Event method do not match the parameters to the WriteEvent method. This may cause the event to be displayed incorrectly.
09:46:11 Discovered:  System.Linq.Expressions.Tests
09:46:11             at System.Diagnostics.Tests.DiagnosticSourceEventSourceBridgeTests.TestNulls()
```

vancem commented:
@stephentoub On the Ubuntu failure above

There does seem to be a race that happens to be triggered by the NullTest.   I am tracking the issue with #6872.   Disabling concurrency does seem to things not fail on my machine (we will see about Ubuntu).   

The error message about mismatched event seems bogus and also needs investigation.

stephentoub commented:
> The error message about mismatched event seems bogus and also needs investigation.

Note that this is likely happening on Windows as well but isn't getting noticed.  On our unix runs we set an environment variable that causes libcoreclr's PAL layer to route OutputDebugString to stderr.  EventSource is likely calling Debug.WriteLine or something like that as part of reporting out-of-band errors or something, and that warning is then showing up in the log because it got routed to stderr.

weshaggard commented on src/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj:
```diff
@@ -7,23 +7,44 @@
+    <NuGetTargetMoniker Condition="'$(TargetGroup)' == ''">.NETStandard,Version=v1.3</NuGetTargetMoniker>
   </PropertyGroup>
-  <ItemGroup Condition="'$(PackageTargetFramework)' == ''">
+
```
nit: Multiple empty lines

weshaggard commented on src/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj:
```diff
@@ -7,23 +7,44 @@
   </ItemGroup>
+
+  <PropertyGroup Condition="'$(TargetGroup)' == 'netstandard1.1'">
+      <DefineConstants>NO_EVENTSOURCE_COMPLEX_TYPE_SUPPORT</DefineConstants>
```
Should be `<DefineConstants>$(DefinedConstants);NO_EVENTSOURCE_COMPLEX_TYPE_SUPPORT</DefineConstants>` otherwise you are throwing away any other defines.

vancem replied:
fixed.

weshaggard commented on src/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj:
```diff
@@ -7,23 +7,44 @@
+
   <!-- Default configurations to help VS understand the configurations -->
-  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
-    <NoWarn>0420</NoWarn>
```
While you might now need the NoWarn you should specific the different configurations here so that you can build the different configurations in VS.

vancem replied:
Fixed (the fact this was lost was not intended (seems to be a merge issue?).

vancem commented:
@dotnet-bot  pleasetestme

I made inconsequential changes and a very unrelated test (System.Net) failed with a timeout.   Trying again as it presumably is a non-deterministic failure.

vancem commented:
@dotnet-bot test this please

vancem commented:
@stephentoub, @weshaggard   All tests pass and I believe that I have addressed all feedback.  I would like to get this in today before we fork for the release.

weshaggard commented on src/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj:
```diff
@@ -7,23 +7,50 @@
     <NoWarn>0420</NoWarn>
   </PropertyGroup>
+
+  <!-- Default configurations to help VS understand the configurations -->
```
nit: this commit got moved, it should be above the configurations and doesn't make since in its current location.

weshaggard commented:
One nit but otherwise LGTM (although I must admit I only reviewed the project stuff and didn't dive deep into the actual code).
